File: src/p0001/erlang/solution.erl
-spec two_sum(Nums :: [integer()], Target :: integer()) -> [integer()].
two_sum(Nums, Target) ->
    two_sum(Nums, 0, Target, #{}).

two_sum([Y | Rest], J, Target, Dict) ->
    X = Target - Y,
    case Dict of
        #{X := I} ->
            [I, J];
        _ ->
            two_sum(Rest, J + 1, Target, Dict#{Y => J})
    end;
two_sum(_, _, _, _) -> [].

File: src/p0001/elixir/solution.ex
defmodule Solution do
  @spec two_sum(nums :: [integer], target :: integer) :: [integer]
  def two_sum(nums, target) do
    two_sum(nums, 0, target, %{})
  end

  def two_sum([y | rest], j, target, dict) do
    x = target - y
    case dict do
      %{^x => i} ->
        [i, j]
      _ ->
        two_sum(rest, j + 1, target, Map.put(dict, y, j))
    end
  end
  def two_sum(_, _, _, _) do
    []
  end
end
